src/prompt/functions/ai-chain.ts

```diff
@@ -0,0 +1,174 @@
+ *
+ * The chain ends when a non-tool call is returned, and the final result can
+ * optionally be validated against a Zod schema, which defaults to a `string`.
+ *
+ * To prevent possible infinite loops, the chain will throw an error if it
+ * exceeds `maxCalls` (`maxCalls` is expected to be >= `maxRetries`).
+ *
+ * Note that `createAIChain` takes in a `functions` array of `AIFunction`
+ * objects created by `createAIFunction`, as the two utility functions are
+ * meant to be used together.
+ */
+export function createAIChain<
+  Params extends Prompt.ChainParams = void,
+  Result extends Prompt.ChainResult = string
+>(
+  {
+    chatModel,
+    schema,
+    params,
+    prompt,
+    functions,
+    maxCalls = 5,
+    maxRetries = 3,
+    toolCallConcurrency = 8,
+  }: {
+    chatModel: Model.Chat.Model;
+    schema?: ZodType<Result>;
+    params?: Omit<
+      Prettify<Model.Chat.Run & Model.Chat.Config>,
+      'functions' | 'tools' | 'function_call' | 'tool_choice'
+    >;
+    prompt?: Prompt.Template<Params>;
+    functions?: Prompt.AIFunction[];
+    maxCalls?: number;
+    maxRetries?: number;
+    toolCallConcurrency?: number;
+  },
+  context?: Model.Ctx
+): Prompt.Chain<Params, Result> {
+  return async (promptArgs: Params): Promise<Result> => {
+    if (!params?.messages && !prompt) {
+      throw new Error('Either params.messages or prompt must be provided');
+    }
+
+    if (params?.messages && prompt) {
+      throw new Error('Only one of params.messages or prompt may be provided');
+    }
+
+    const messages =
+      params?.messages ?? (await Promise.resolve(prompt!(promptArgs)));
+    const tools = functions?.map((func) => ({
+      type: 'function' as const,
+      function: func.spec,
+    }));
+
+    let numCalls = 0;
+    let numErrors = 0;
+
+    while (numCalls < maxCalls) {
+      ++numCalls;
+      const response = await chatModel.run(
+        {
+          ...params,
+          messages,
+          tools,
+        },
+        context
+      );
+
+      const { message } = response;
+      try {
+        if (Msg.isToolCall(message)) {
+          messages.push(message);
+
+          if (!functions) {
+            throw new AbortError('No functions provided to handle tool call');
+          }
+
+          await pMap(
+            message.tool_calls,
+            async (toolCall) => {
+              const func = functions?.find(
+                (func) => func.spec.name === toolCall.function.name
+              );
+
+              if (!func) {
+                throw new Error(
+                  `No function found with name ${toolCall.function.name}`
+                );
+              }
+
+              // TODO: ideally we'd differentiate between tool argument validation
+              // errors versus errors in the function implementation
+              const result = await func(toolCall.function.arguments);
+
+              const toolResult = Msg.toolResult(
+                JSON.stringify(result, null, 2),
```

Comment:
i thought we agreed not to prettify json sent to gpt?

Reply:
yep :) I added a `stringifyForModel` that gets implicitly called when you pass the function results to `Msg.toolResult` and `Msg.funcResult`, but I think I was pushing that change while you were reviewing

Reply:
oh i dont think we should be doing that as a default. its not something we have extensively tested and are 100% certain of, so we shouldnt make it impossible to do.

Reply:
I updated this s.t. if you pass a string to `toolResult` or `funcResult`, it won't stringify it again